src/misc/umap.c

```diff
@@ -0,0 +1,846 @@
+ * feed a k-nearest neighbor (either computed exactly or approximated) similarity graph
+ * with some additional cutoff to exclude "quasi-neighbors" that lie beyond a certain
+ * distance (e.g. correlation less than 0.2).
+ *
+ * Therefore, if you are trying to use this function to embed high-dimensional vectors,
+ * the steps are:
+ *
+ * 1. Compute a sparse similarity graph (either exact or approximate) from your vectors,
+ *    weighted or unweighted. If unsure, compute a knn.
+ * 2. If you keep the weights, convert them into distances or store them as a "weight"
+ *    edge attribute and use a null pointer for the distances. If using similarity
+ *    weights instead of distances, make sure they do not exceed 1.
+ * 3. Feed the graph (and distances, if you have them) into this function.
+ *
+ * Note: Step 1 above involves deciding if two high-dimensional vectors "look similar"
+ *       which, because of the curse of dimensionality, is in many cases a highly
+ *       subjective and potentially controversial operation: thread with care and at
+ *       your own risk. Two high-dimensional vectors might look similar or extremely
+ *       different depending on the point of view/angle, and there are a lot of
+ *       viewpoints when the dimensionality ramps up.
+ *
+ * </para><para>
+ * References:
+ *
+ * </para><para>
+ * Leland McInnes, John Healy, and James Melville, arXiv:1802.03426v3
+ *
+ * \param graph Pointer to the similarity graph to find a layout for (i.e. to embed). Weights in
+ *   this graph object are currently ignored.
+ * \param distances Pointer to a vector of "distances" between vertices. Similarity graphs for
+ *   UMAP are often originally meant in terms of similarity weights (e.g. correlation between
+ *   high-dimensional vectors) and converted into distances by crude dist := 1 - corr. That is
+ *   fine here too. If this argument is an empty pointer (NULL), connected vertices are assumed
+ *   to be very similar.
+ * \param layout Pointer to the n x 2 matrix where the layout coordinates will be stored. Only 2D
+ *   embedding are currently supported, as they are by far more common than any higher dimensions.
+ * \param min_dist A fudge parameter that decides how close two unconnected vertices can be in the
+ *   embedding before feeling a repulsive force. Typically, 0.01 is a good number. If this is
+ *   negative or zero, 0.01 is assumed.
+ * \param epochs Number of iterations of the main stochastic gradient descent loop on the
+ *   cross-entropy. If negative or zero, 500 epochs are used if the graph is the graph is small
+ *   (less than 50k edges), 50 epochs are used for larger graphs.
+ * \param sampling_prob The fraction of vertices moved at each iteration of the stochastic gradient descent (epoch). At fixed number of epochs, a higher fraction makes the algorithm slower. Vice versa, a too low number will converge very slowly, possibly too slowly. If negative of zero, a value of 0.3 is assumed.
```

Comment:
Wrap this line. Throw an error for invalid values, do not fall back to defaults.